src/lib/arch/riscv/Instruction_decode.cc

```diff
@@ -0,0 +1,254 @@
+        } else {
+          operandsPending++;
+        }
+
+        sourceRegisterCount++;
+      } else {
+        if (csRegToRegister(op.reg) != Instruction::ZERO_REGISTER) {
+          destinationRegisters[destinationRegisterCount] =
+              csRegToRegister(op.reg);
+
+          destinationRegisterCount++;
+        }
+      }
+    }
+
+    // For all instructions, every register operand after the first is a source
+    // register
+    else if (i > 0 && op.type == RISCV_OP_REG) {
+      //  Second or third operand
+      sourceRegisters[sourceRegisterCount] = csRegToRegister(op.reg);
+
+      if (sourceRegisters[sourceRegisterCount] == Instruction::ZERO_REGISTER) {
+        // Catch zero register references and pre-complete those operands
+        operands[sourceRegisterCount] = RegisterValue(0, 8);
+      } else {
+        operandsPending++;
+      }
+
+      sourceRegisterCount++;
+    }
+
+    // First operand is never MEM type, only check after the first. If register
+    // contains memory address, extract reg number from capstone object
+    else if (i > 0 && op.type == RISCV_OP_MEM) {
+      //  Memory operand
+      sourceRegisters[sourceRegisterCount] = csRegToRegister(op.mem.base);
+      sourceRegisterCount++;
+      operandsPending++;
+    }
+  }
+
+  if ((Opcode::RISCV_SLL <= metadata.opcode &&
+       metadata.opcode <= Opcode::RISCV_SLLW) ||
+      (Opcode::RISCV_SRA <= metadata.opcode &&
+       metadata.opcode <= Opcode::RISCV_SRAW) ||
+      (Opcode::RISCV_SRL <= metadata.opcode &&
+       metadata.opcode <= Opcode::RISCV_SRLW)) {
+    // Shift instructions
+    isShift_ = true;
+  }
+
+  if ((Opcode::RISCV_XOR <= metadata.opcode &&
+       metadata.opcode <= Opcode::RISCV_XORI) ||
+      (Opcode::RISCV_OR <= metadata.opcode &&
+       metadata.opcode <= Opcode::RISCV_ORI) ||
+      (Opcode::RISCV_AND <= metadata.opcode &&
+       metadata.opcode <= Opcode::RISCV_ANDI)) {
+    // Logical instructions
+    isLogical_ = true;
+  }
+
+  if ((Opcode::RISCV_SLT <= metadata.opcode &&
+       metadata.opcode <= Opcode::RISCV_SLTU)) {
+    // Compare instructions
+    isCompare_ = true;
+  }
+
+  // Set branch type
+  switch (metadata.opcode) {
+    case Opcode::RISCV_BEQ:
+    case Opcode::RISCV_BNE:
+    case Opcode::RISCV_BLT:
+    case Opcode::RISCV_BLTU:
+    case Opcode::RISCV_BGE:
+    case Opcode::RISCV_BGEU:
+      branchType_ = BranchType::Conditional;
+      knownTarget_ = instructionAddress_ + metadata.operands[2].imm;
+      break;
+    case Opcode::RISCV_JAL:
+    case Opcode::RISCV_JALR:
+      branchType_ = BranchType::Unconditional;
+      knownTarget_ = instructionAddress_ + metadata.operands[1].imm;
+      break;
+  }
+}
+
+void Instruction::nyi() {
+  exceptionEncountered_ = true;
+  exception_ = InstructionException::EncodingNotYetImplemented;
+}
+void Instruction::unallocated() {
+  exceptionEncountered_ = true;
+  exception_ = InstructionException::EncodingUnallocated;
+}
```

Comment:
why are these functions defined here rather than execute / address / metadata?

Reply:
This follows the AArch64 implementation

Reply:
Please test if they are required in this class, or if another class depends on them. If appropriate please move them to the correct class

Reply:
Removed in both AArch64 and RISC-V implementations